hmmlearn/tests/test_gmm_hmm_new.py

```diff
@@ -0,0 +1,165 @@
+import numpy as np
+from hmmlearn.hmm import GMMHMM
+
+import pytest
+from sklearn.mixture.gmm import sample_gaussian
+
+from sklearn.preprocessing import normalize
+
+from . import normalized
+
+
+def sample_from_parallelepiped(low, high, n_samples, random_state):
+    (n_features,) = low.shape
+    X = np.zeros((n_samples, n_features))
+    for i in range(n_features):
+        X[:, i] = random_state.uniform(low[i], high[i], n_samples)
+    return X
+
+
+def prep_params(n_comps, n_mix, n_features, covar_type,
+                low, high, random_state):
+    # the idea is to generate ``n_comps`` bounding boxes and then
+    # generate ``n_mix`` mixture means in each of them
+
+    dim_lims = np.zeros((n_comps + 1, n_features))
+
+    # this generates a sequence of coordinates, which are then used as
+    # vertices of bounding boxes for mixtures
+    dim_lims[1:] = np.cumsum(
+        random_state.uniform(low, high, (n_comps, n_features)), axis=0
+    )
+
+    means = np.zeros((n_comps, n_mix, n_features))
+    for i, (left, right) in enumerate(zip(dim_lims, dim_lims[1:])):
+        means[i] = sample_from_parallelepiped(left, right, n_mix,
+                                              random_state)
+
+    startprob = np.zeros(n_comps)
+    startprob[0] = 1
+
+    transmat = random_state.uniform(size=(n_comps, n_comps))
+    transmat /= np.sum(transmat, axis=1)[:, np.newaxis]
+
+    if covar_type == "spherical":
+        covs = random_state.uniform(0.1, 5, size=(n_comps, n_mix))
+    elif covar_type == "diag":
+        covs = random_state.uniform(0.1, 5, size=(n_comps, n_mix, n_features))
+    elif covar_type == "tied":
+        covs = np.zeros((n_comps, n_features, n_features))
+        for i in range(n_comps):
+            low = random_state.uniform(-2, 2, (n_features, n_features))
+            covs[i] = np.dot(low.T, low)
+    elif covar_type == "full":
+        covs = np.zeros((n_comps, n_mix, n_features, n_features))
+        for i in range(n_comps):
+            for j in range(n_mix):
+                low = random_state.uniform(-2, 2,
+                                           size=(n_features, n_features))
+                covs[i, j] = np.dot(low.T, low)
+
+    weights = random_state.uniform(size=(n_comps, n_mix))
+    weights /= np.sum(weights, axis=1)[:, np.newaxis]
+
+    return covs, means, startprob, transmat, weights
+
+
+class GMMHMMTestMixin(object):
+    def setup_method(self, method):
+        self.prng = np.random.RandomState(14)
+        self.n_components = 3
+        self.n_mix = 2
+        self.n_features = 2
+
+        low, high = 10, 15
+        (self.covars, self.means,
+         self.startprob, self.transmat, self.weights) = prep_params(
+            self.n_components, self.n_mix, self.n_features,
+            self.covariance_type, low, high, self.prng
+        )
+        
+        self.h = GMMHMM(n_components=self.n_components, n_mix=self.n_mix,
```

Comment:
I did it because previous version of tests for GMMHMM and other tests use this pattern. But if we're going to use `pytest.mark.parametrize` and don't use classes, I don't quite see, how are we going to do setup for each function. I don't think it's possible to parametrize setup functions.